File: apps/backend/agent-api/api/routers/flashcards.py
from fastapi import APIRouter, HTTPException

from api.services import flashcard_service

router = APIRouter(prefix="/api/spaces", tags=["flashcards"])


@router.post("/{space_slug}/topics/{topic_slug}/flashcards")
def generate_flashcards(space_slug: str, topic_slug: str):
    try:
        flashcards = flashcard_service.generate_topic_flashcards(space_slug, topic_slug)
        return {"flashcards": flashcards}
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/{space_slug}/topics/{topic_slug}/flashcards")
def get_flashcards(space_slug: str, topic_slug: str):
    try:
        flashcards = flashcard_service.get_topic_flashcards(space_slug, topic_slug)
        return {"flashcards": flashcards}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.delete("/{space_slug}/topics/{topic_slug}/flashcards")
def delete_flashcards(space_slug: str, topic_slug: str):
    try:
        flashcard_service.delete_topic_flashcards(space_slug, topic_slug)
        return {"message": "Flashcards deleted"}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

File: apps/frontend/web/src/api/flashcards.ts
const API_BASE = 'http://localhost:8000'

export interface FlashcardContent {
  html: string
}

export interface Flashcard {
  id: number
  front: FlashcardContent
  back: FlashcardContent
}

export interface FlashcardsResponse {
  flashcards: Flashcard[]
}

async function fetchApi<T>(url: string, options?: RequestInit): Promise<T> {
  const response = await fetch(`${API_BASE}${url}`, {
    ...options,
    headers: {
      'Content-Type': 'application/json',
      ...options?.headers,
    },
  })

  if (!response.ok) {
    const error = await response.json().catch(() => ({ detail: 'Unknown error' }))
    throw new Error(error.detail || 'API error')
  }

  return response.json()
}

export async function generateFlashcards(spaceSlug: string, topicSlug: string): Promise<FlashcardsResponse> {
  return fetchApi<FlashcardsResponse>(`/api/spaces/${spaceSlug}/topics/${topicSlug}/flashcards`, {
    method: 'POST',
  })
}

export async function getFlashcards(spaceSlug: string, topicSlug: string): Promise<FlashcardsResponse> {
  return fetchApi<FlashcardsResponse>(`/api/spaces/${spaceSlug}/topics/${topicSlug}/flashcards`)
}

export async function deleteFlashcards(spaceSlug: string, topicSlug: string): Promise<{ message: string }> {
  return fetchApi<{ message: string }>(`/api/spaces/${spaceSlug}/topics/${topicSlug}/flashcards`, {
    method: 'DELETE',
  })
}
